tests/batch_2/SingleOwnerTonWallet.admin.spec.ts

```diff
@@ -0,0 +1,344 @@
+            from: admin.address,
+            to: singleOwnerTonWallet.address,
+            success: true,
+            op: Number(SingleOwnerTonWalletOpcode.upgrade_code),
+        });
+        const newVersion = await singleOwnerTonWallet.getVersion();
+        expect(newVersion).toEqual(NEW_VERSION);
+
+        // excess
+        expect(sendUpgradeCodeResult.transactions).toHaveTransaction({
+            from: singleOwnerTonWallet.address,
+            to: admin.address,
+            success: true,
+            op: Number(SingleOwnerTonWalletOpcode.excesses),
+        });
+
+        const refundTx = flattenTransaction(
+            findTransactionRequired(sendUpgradeCodeResult.transactions, {
+                from: singleOwnerTonWallet.address,
+                to: admin.address,
+                success: true,
+                op: Number(SingleOwnerTonWalletOpcode.excesses),
+            }),
+        );
+
+        const refundAmount = refundTx.value!;
+        const approxOtherFeesLowerBound = toNano('0.0008');
+        const approxOtherFeesUpperBound = toNano('0.0019');
+        const upgradeInStorageFee = toNano(`0.002`);
+
+        expect(refundAmount).toBeLessThan(totalFees - upgradeInStorageFee - approxOtherFeesLowerBound);
+        expect(refundAmount).toBeGreaterThan(totalFees - upgradeInStorageFee - approxOtherFeesUpperBound);
+
+        cleanupTestContract();
+    });
+
+    it(`should refuse to upgrade sotw code if the sender is not the admin`, async () => {
+        const NEW_VERSION = 99999999999999n;
+
+        const { code: SingleOwnerTonWalletUpgradeTestCode, cleanupTestContract } =
+            await createSingleOwnerTonWalletUpgradeTestContractCode(NEW_VERSION);
+
+        // Test validating the existing data after upgrade
+        const prevVersion = await singleOwnerTonWallet.getVersion();
+
+        const randomUser = await blockchain.treasury('randomUser');
+
+        const sendUpgradeCodeResult = await singleOwnerTonWallet.sendUpgradeCode(
+            randomUser.getSender(),
+            toNano('0.013'),
+            {
+                newCode: SingleOwnerTonWalletUpgradeTestCode,
+            },
+        );
+
+        expect(sendUpgradeCodeResult.transactions).toHaveTransaction({
+            from: randomUser.address,
+            to: singleOwnerTonWallet.address,
+            success: false,
+            op: Number(SingleOwnerTonWalletOpcode.upgrade_code),
+        });
+
+        // Check the version is not updated
+        const currentVersion = await singleOwnerTonWallet.getVersion();
+        expect(currentVersion).toEqual(prevVersion);
+        cleanupTestContract();
+    });
+});
+
+async function createSingleOwnerTonWalletUpgradeTestContractCode(version: bigint, dumbOutput?: number) {
+    const BASE_RELATIVE__PATH = '../../';
+    // Preparing the new code for upgrade contract
+    const singleOwnerTonWalletUpgradeTestPath = path.join(
+        __dirname,
+        BASE_RELATIVE__PATH,
```

Comment:
```suggestion
        '..', '..',
```
I am not sure if the current code will run on Windows (if anyone in the future wants to run tests on Windows)

Reply:
commit 6d5a0edcfac2bacb3f6906566c2310b0b899d169